scrape/scrape.go

```diff
@@ -643,28 +646,44 @@ func newScrapeCache() *scrapeCache {
 	}
 }
 
-func (c *scrapeCache) iterDone() {
-	// All caches may grow over time through series churn
-	// or multiple string representations of the same metric. Clean up entries
-	// that haven't appeared in the last scrape.
-	for s, e := range c.series {
-		if c.iter-e.lastIter > 2 {
-			delete(c.series, s)
+func (c *scrapeCache) iterDone(flushCache bool) {
+	c.metaMtx.Lock()
+	count := len(c.series) + len(c.droppedSeries) + len(c.metadata)
+	c.metaMtx.Unlock()
+
+	if flushCache {
+		c.successfulCount = count
+	} else if count > c.successfulCount*2+100 {
```

Comment:
My understanding is that this is a heuristic, that aims for tolerating targets that drop and replace all metrics in a consecutive scrape plus the metadata (which should be somewhat consistent in either case). At the very least this needs a comment describing the heuristic, but I'd also like us to consider different heuristics. For example as I said metadata is probably irrelevant, at least in terms of churn. So maybe just x2 or x3, without metadata is simpler and should allow us to catch close to all churn situations. Please correct me if I misunderstood any of this, which may very well have happened.

Reply:
This is more that if there's a small bit of churn don't lose anything. The 100 is to somewhat handle the case where this happens before there's been a successful scrape.

Metadata churn is a possibility. We've never seen it, but may aswell protect against it while we're at it.

Reply:
I’d like us to at least document the heuristic and maybe explain that flushing the cache on initial iter is ok (as successfulCount will still be 0 and count is likely to be higher than 100)

Reply:
Done, and bumped it to 1k.